<h1>Chapter 2 | Data Exercise #4 | <code>world-bank-immunization</code>: Data Preparation</h1>

<p>4. Consider the <code>world-bank-immunization</code> dataset based on World Bank data.</p>
<p>Assignments</p>
<ul>
    <li>Generate a new variable for the growth rate of GDP per capita, in percentage terms.</li>
    
    

$(gdppc_{it}/gdpcc_{i,t-1})*100$

<ul>
    <li>Add this variable to the long and wide format of the data.</li>
    <li>Add the variabe to Tables 2.4 and 2.5.</li>
</ul>

<h2><b>1.</b> Load the data</h2>

In [38]:
import os
import sys
import warnings
import pandas as pd

warnings.filterwarnings("ignore")

In [39]:
# Current script folder
current_path = os.getcwd()
dirname = current_path.split("da_data_exercises")[0]

# Get location folders
data_in = f"{dirname}da_data_repo/worldbank-immunization/clean/"
data_out = f"{dirname}da_data_exercises/ch02-preparing_data_for_analysis/data"
output = f"{dirname}da_data_exercises/ch02-preparing_data_for_analysis/data/output/"
func = f"{dirname}da_case_studies/ch00-tech_prep/"
sys.path.append(func)

In [40]:
# Import the prewritten helper functions
from py_helper_functions import *

In [41]:
data = pd.read_csv(f"{data_in}worldbank-immunization-panel.csv")

<p>Let's take a brief look at the data:</p>

In [42]:
data.head()

year    c countryname countrycode          pop  mort   surv  imm  \
0  1998  ARG   Argentina         ARG    36.063459  21.5  97.85   95   
1  1998  AUS   Australia         AUS    18.711000   6.4  99.36   82   
2  1998  BRA      Brazil         BRA   169.785250  39.5  96.05   95   
3  1998  CHN       China         CHN  1241.935000  41.4  95.86   83   
4  1998  FRA      France         FRA    60.186288   5.6  99.44   82   

          gdppc    lngdppc  hexp  
0  15973.067259   9.678659   NaN  
1  33174.499840  10.409537   NaN  
2  11193.381175   9.323078   NaN  
3   3211.623211   8.074532   NaN  
4  32679.688245  10.394509   NaN

In [43]:
data.tail()

year    c    countryname countrycode        pop  mort   surv  imm  \
3802  2017  VEN  Venezuela, RB         VEN  29.390409  30.9  96.91   96   
3803  2017  VNM        Vietnam         VNM  94.596642  20.9  97.91   97   
3804  2017  YEM    Yemen, Rep.         YEM  27.834821  55.4  94.46   65   
3805  2017  ZMB         Zambia         ZMB  16.853688  60.0  94.00   96   
3806  2017  ZWE       Zimbabwe         ZWE  14.236745  50.3  94.97   90   

            gdppc   lngdppc  hexp  
3802          NaN       NaN   NaN  
3803  6233.485045  8.737691   NaN  
3804  2404.422370  7.785065   NaN  
3805  3717.667166  8.220852   NaN  
3806  2568.410072  7.851042   NaN

In [44]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3807 entries, 0 to 3806
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   year         3807 non-null   int64  
 1   c            3807 non-null   object 
 2   countryname  3807 non-null   object 
 3   countrycode  3807 non-null   object 
 4   pop          3807 non-null   float64
 5   mort         3807 non-null   float64
 6   surv         3807 non-null   float64
 7   imm          3807 non-null   int64  
 8   gdppc        3642 non-null   float64
 9   lngdppc      3642 non-null   float64
 10  hexp         3165 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 327.3+ KB


In [45]:
data.shape

(3807, 11)

<h3>1.1. <b>Task #1 </b>| Generate a new variable for the growth rate of GDP per capita, in percentage terms</h3>
<p>Basically, we have to:</p>
<ul>
    <li>Divide <code>gdppc</code> by the value registered in the previous year</li>
    <li>Multiply by 100</li>
</ul>
<p>We can use <code>pct_change()</code> to get the percentage change from one year to the next:</p>

In [46]:
# Sort the dataframe by country and year code
data = data.sort_values(["countrycode", "year"]).reset_index(drop=True)

In [47]:
# Calculate the gdppc growth using pct_change
data["gdppc_growth"] = data.groupby("countrycode")["gdppc"].pct_change()

In [48]:
data.head(20)

year    c  countryname countrycode        pop   mort   surv  imm  \
0   1998  AFG  Afghanistan         AFG  19.737765  135.8  86.42   31   
1   1999  AFG  Afghanistan         AFG  20.170844  132.3  86.77   31   
2   2000  AFG  Afghanistan         AFG  20.779953  128.8  87.12   27   
3   2001  AFG  Afghanistan         AFG  21.606988  125.3  87.47   37   
4   2002  AFG  Afghanistan         AFG  22.600770  121.7  87.83   35   
5   2003  AFG  Afghanistan         AFG  23.680871  117.9  88.21   39   
6   2004  AFG  Afghanistan         AFG  24.726684  114.1  88.59   48   
7   2005  AFG  Afghanistan         AFG  25.654277  110.1  88.99   50   
8   2006  AFG  Afghanistan         AFG  26.433049  106.1  89.39   53   
9   2007  AFG  Afghanistan         AFG  27.100536  102.0  89.80   55   
10  2008  AFG  Afghanistan         AFG  27.722276   97.9  90.21   59   
11  2009  AFG  Afghanistan         AFG  28.394813   93.9  90.61   60   
12  2010  AFG  Afghanistan         AFG  29.185507   90.0  91.00   62   
13  2011  AFG  Afghanistan         AFG  30.117413   86.2  91.38   64   
14  2012  AFG  Afghanistan         AFG  31.161376   82.6  91.74   59   
15  2013  AFG  Afghanistan         AFG  32.269589   79.2  92.08   57   
16  2014  AFG  Afghanistan         AFG  33.370794   76.0  92.40   60   
17  2015  AFG  Afghanistan         AFG  34.413603   73.1  92.69   62   
18  2016  AFG  Afghanistan         AFG  35.383128   70.4  92.96   62   
19  2017  AFG  Afghanistan         AFG  36.296400   67.9  93.21   62   

          gdppc   lngdppc       hexp  gdppc_growth  
0           NaN       NaN        NaN           NaN  
1           NaN       NaN        NaN           NaN  
2           NaN       NaN        NaN           NaN  
3           NaN       NaN        NaN           NaN  
4   1016.245409  6.923870   9.443391           NaN  
5   1055.557459  6.961824   8.941259      0.038684  
6   1025.208245  6.932651   9.808473     -0.028752  
7   1099.104568  7.002251   9.948289      0.072079  
8   1123.871323  7.024535  10.622766      0.022534  
9   1247.753118  7.129100   9.904674      0.110228  
10  1267.644939  7.144916  10.256496      0.015942  
11  1502.354073  7.314789   9.818487      0.185154  
12  1671.581238  7.421525   8.569671      0.112641  
13  1626.764793  7.394349   8.561907     -0.026811  
14  1772.764974  7.480296   7.897168      0.089749  
15  1807.762344  7.499845   8.805964      0.019742  
16  1795.735834  7.493170   9.528878     -0.006653  
17  1766.593077  7.476808  10.093958     -0.016229  
18  1757.023490  7.471376  10.198677     -0.005417  
19  1758.465636  7.472197        NaN      0.000821

In [49]:
data[["year","countryname","gdppc", "gdppc_growth"]].head(30)

year  countryname        gdppc  gdppc_growth
0   1998  Afghanistan          NaN           NaN
1   1999  Afghanistan          NaN           NaN
2   2000  Afghanistan          NaN           NaN
3   2001  Afghanistan          NaN           NaN
4   2002  Afghanistan  1016.245409           NaN
5   2003  Afghanistan  1055.557459      0.038684
6   2004  Afghanistan  1025.208245     -0.028752
7   2005  Afghanistan  1099.104568      0.072079
8   2006  Afghanistan  1123.871323      0.022534
9   2007  Afghanistan  1247.753118      0.110228
10  2008  Afghanistan  1267.644939      0.015942
11  2009  Afghanistan  1502.354073      0.185154
12  2010  Afghanistan  1671.581238      0.112641
13  2011  Afghanistan  1626.764793     -0.026811
14  2012  Afghanistan  1772.764974      0.089749
15  2013  Afghanistan  1807.762344      0.019742
16  2014  Afghanistan  1795.735834     -0.006653
17  2015  Afghanistan  1766.593077     -0.016229
18  2016  Afghanistan  1757.023490     -0.005417
19  2017  Afghanistan  1758.465636      0.000821
20  1998       Angola  3945.594930           NaN
21  1999       Angola  3902.774196     -0.010853
22  2000       Angola  3892.316925     -0.002679
23  2001       Angola  3924.316196      0.008221
24  2002       Angola  4314.540938      0.099438
25  2003       Angola  4295.908569     -0.004319
26  2004       Angola  4604.657044      0.071870
27  2005       Angola  5112.589202      0.110308
28  2006       Angola  5500.242536      0.075823
29  2007       Angola  6044.217156      0.098900

<p>Great! Likewise, we can also use the <code>.shift()</code> to get the previous year's <code>gdppc</code> and then divide the current year's <code>gdppc</code> by the sifted value and muliply it by 1000 to get the percentage change. Let's try it in a new variable:</p>

In [50]:
data["gdppc_pct_change"] = data.groupby("countrycode")["gdppc"].apply(lambda x: x / x.shift(1) - 1)

In [51]:
data.head(10)

year    c  countryname countrycode        pop   mort   surv  imm  \
0  1998  AFG  Afghanistan         AFG  19.737765  135.8  86.42   31   
1  1999  AFG  Afghanistan         AFG  20.170844  132.3  86.77   31   
2  2000  AFG  Afghanistan         AFG  20.779953  128.8  87.12   27   
3  2001  AFG  Afghanistan         AFG  21.606988  125.3  87.47   37   
4  2002  AFG  Afghanistan         AFG  22.600770  121.7  87.83   35   
5  2003  AFG  Afghanistan         AFG  23.680871  117.9  88.21   39   
6  2004  AFG  Afghanistan         AFG  24.726684  114.1  88.59   48   
7  2005  AFG  Afghanistan         AFG  25.654277  110.1  88.99   50   
8  2006  AFG  Afghanistan         AFG  26.433049  106.1  89.39   53   
9  2007  AFG  Afghanistan         AFG  27.100536  102.0  89.80   55   

         gdppc   lngdppc       hexp  gdppc_growth  gdppc_pct_change  
0          NaN       NaN        NaN           NaN               NaN  
1          NaN       NaN        NaN           NaN               NaN  
2          NaN       NaN        NaN           NaN               NaN  
3          NaN       NaN        NaN           NaN               NaN  
4  1016.245409  6.923870   9.443391           NaN               NaN  
5  1055.557459  6.961824   8.941259      0.038684          0.038684  
6  1025.208245  6.932651   9.808473     -0.028752         -0.028752  
7  1099.104568  7.002251   9.948289      0.072079          0.072079  
8  1123.871323  7.024535  10.622766      0.022534          0.022534  
9  1247.753118  7.129100   9.904674      0.110228          0.110228

<p>It worked like a charm! In this case, we grouped the DataFrame by <code>countrycode</code> and applied a lambda function to <code>gdppc</code>.<p>
<p>As a reminder, <code>lambda</code> is a keyword used to define an <b>anonymous function</b> that can take can take any number of arguments but can have only <b>one expression</b>. Here, we used it taking a Pandas Series as its <code>x</code> argument, that is, <code>gdppc</code>, and a new Series object will be returned by applying the following operations:</p>
<ol>
    <li><code>x.shift(1)</code> shifts the values in the Series <code>x</code> down by <b>one row</b>, effectling aligning the current row with the previous row.</li>
    <li><code>x / shift(1) - 1</code> calculates the difference change between the current row and the previous row.</li>
    <li><code>(x / shift(1) - 1) * 100</code> multiplies the result by 100 to convert the percentage change to a percentage growth rate.</li>
</ol>
<p>Finally, with <code>.apply()</code>, we passed this anonymous function as an argument inside it to the Pandas Series, creating the percentage growth rate of GDP per capita for each country and year.</p>
<p>Ok, so we already have added the <code>gdppc_pct_change</code> to a long format. How could we achieve the same result in a wide format data?</p>
<p>We can do this by trying to reproduce table 2.5.</p>

In [52]:
data_wide = data.set_index(["countryname", "year"]).unstack("year")

In [53]:
data_wide

c                                               ...  \
year          1998 1999 2000 2001 2002 2003 2004 2005 2006 2007  ...   
countryname                                                      ...   
Afghanistan    AFG  AFG  AFG  AFG  AFG  AFG  AFG  AFG  AFG  AFG  ...   
Albania        ALB  ALB  ALB  ALB  ALB  ALB  ALB  ALB  ALB  ALB  ...   
Algeria        DZA  DZA  DZA  DZA  DZA  DZA  DZA  DZA  DZA  DZA  ...   
Andorra        AND  AND  AND  AND  AND  AND  AND  AND  AND  AND  ...   
Angola         AGO  AGO  AGO  AGO  AGO  AGO  AGO  AGO  AGO  AGO  ...   
...            ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
Venezuela, RB  VEN  VEN  VEN  VEN  VEN  VEN  VEN  VEN  VEN  VEN  ...   
Vietnam        VNM  VNM  VNM  VNM  VNM  VNM  VNM  VNM  VNM  VNM  ...   
Yemen, Rep.    YEM  YEM  YEM  YEM  YEM  YEM  YEM  YEM  YEM  YEM  ...   
Zambia         ZMB  ZMB  ZMB  ZMB  ZMB  ZMB  ZMB  ZMB  ZMB  ZMB  ...   
Zimbabwe       ZWE  ZWE  ZWE  ZWE  ZWE  ZWE  ZWE  ZWE  ZWE  ZWE  ...   

              gdppc_pct_change                                          \
year                      2008      2009      2010      2011      2012   
countryname                                                              
Afghanistan           0.015942  0.185154  0.112641 -0.026811  0.089749   
Albania               0.083280  0.040488  0.042230  0.028216  0.015852   
Algeria               0.006990 -0.001030  0.017803  0.009722  0.013771   
Andorra                    NaN       NaN       NaN       NaN       NaN   
Angola                0.071169 -0.028086  0.010792 -0.002208  0.047065   
...                        ...       ...       ...       ...       ...   
Venezuela, RB         0.037989 -0.045669 -0.029053  0.025593  0.039205   
Vietnam               0.046525  0.043706  0.053642  0.051597  0.041563   
Yemen, Rep.           0.007613  0.009863  0.047329 -0.151080 -0.003918   
Zambia                0.048752  0.061904  0.071298  0.024239  0.043104   
Zimbabwe             -0.184911  0.107013  0.180660  0.124529  0.147012   

                                                                 
year               2013      2014      2015      2016      2017  
countryname                                                      
Afghanistan    0.019742 -0.006653 -0.016229 -0.005417  0.000821  
Albania        0.011872  0.019809  0.025281  0.035154  0.039322  
Algeria        0.007300  0.016998  0.016627  0.012025 -0.004566  
Andorra             NaN       NaN       NaN       NaN       NaN  
Angola         0.012921  0.012198 -0.024687 -0.058162 -0.034099  
...                 ...       ...       ...       ...       ...  
Venezuela, RB -0.000898 -0.047311       NaN       NaN       NaN  
Vietnam        0.043172  0.048730  0.055714  0.051201  0.057306  
Yemen, Rep.    0.020143 -0.028030 -0.187991 -0.157527 -0.081949  
Zambia         0.018120  0.014784 -0.001886  0.006873  0.003957  
Zimbabwe       0.001925  0.005962  0.001005 -0.007936  0.031864  

[192 rows x 220 columns]

<p>As we can see, we unstacked the table by using the variable <code>year</code> and set the index using this variable plus <code>countryname</code>. We now have one row per country, which resulted in 192 rows, and 220 columns, as for each column now there is one observation per reported year.</p>
<p>To reproduce Table 2.5, let's filter the required variables and the newly created <code>gdppc_pct_change</code>.

In [54]:
# Keep only required variables and filter variables according to book example
data = data.filter(["countryname", "year", "imm", "gdppc", "gdppc_pct_change"]).loc[
    (data["year"] >= 2015) &
    ((data["countryname"] == "Pakistan") | (data["countryname"] == "India"))
]

In [55]:
data.describe(percentiles=[])

year        imm        gdppc  gdppc_pct_change
count     6.000000   6.000000     6.000000          6.000000
mean   2016.000000  81.500000  5373.962113          0.048794
std       0.894427   6.774954   874.322204          0.019495
min    2015.000000  75.000000  4459.146517          0.025674
50%    2016.000000  81.500000  5257.315866          0.047825
max    2017.000000  88.000000  6516.172620          0.069970

In [56]:
data.filter(["countryname", "year", "imm", "gdppc", "gdppc_pct_change"])

countryname  year  imm        gdppc  gdppc_pct_change
1551       India  2015   87  5743.426497          0.067970
1552       India  2016   88  6145.294595          0.069970
1553       India  2017   88  6516.172620          0.060352
2662    Pakistan  2015   75  4459.146517          0.025674
2663    Pakistan  2016   75  4608.527214          0.033500
2664    Pakistan  2017   76  4771.205236          0.035299

In [57]:
# Table 2.5 + GDP Per Capita % Change
data.sort_values(["countryname", "year"])

countryname  year  imm        gdppc  gdppc_pct_change
1551       India  2015   87  5743.426497          0.067970
1552       India  2016   88  6145.294595          0.069970
1553       India  2017   88  6516.172620          0.060352
2662    Pakistan  2015   75  4459.146517          0.025674
2663    Pakistan  2016   75  4608.527214          0.033500
2664    Pakistan  2017   76  4771.205236          0.035299

<p>We can now unstack this table using <code>.unstack()</code> and define a new index based on <code>countryname</code> and <code>year</code> variables:</p>

In [58]:
data_wide = data.set_index(["countryname", "year"]).unstack("year")
data_wide

imm                  gdppc                            \
year        2015 2016 2017         2015         2016         2017   
countryname                                                         
India         87   88   88  5743.426497  6145.294595  6516.172620   
Pakistan      75   75   76  4459.146517  4608.527214  4771.205236   

            gdppc_pct_change                     
year                    2015     2016      2017  
countryname                                      
India               0.067970  0.06997  0.060352  
Pakistan            0.025674  0.03350  0.035299

<p>Since we genereated a hieararchical MultiIndex, we can reset the column names and generate better names.</p>

In [59]:
data_wide.columns = [x[0] + "_" + str(x[1]) for x in data_wide.columns]

In [60]:
# Table 2.5
data_wide

imm_2015  imm_2016  imm_2017   gdppc_2015   gdppc_2016  \
countryname                                                           
India              87        88        88  5743.426497  6145.294595   
Pakistan           75        75        76  4459.146517  4608.527214   

              gdppc_2017  gdppc_pct_change_2015  gdppc_pct_change_2016  \
countryname                                                              
India        6516.172620               0.067970                0.06997   
Pakistan     4771.205236               0.025674                0.03350   

             gdppc_pct_change_2017  
countryname                         
India                     0.060352  
Pakistan                  0.035299

<p>And that's it for Data Exercise #4!</p>
<hr>